_posts/2018-07-10-react-native-wdio.md

```diff
@@ -0,0 +1,321 @@
+```
+
+**Run the test:**
+Let's run the test using Sauce Labs with the updated configuration.
+
+```console
+wdio wdio.conf.js
+```
+
+Check the [Sauce Labs Dashboard](https://saucelabs.com/beta/dashboard/tests) to make sure the test ran successfully. The WebDriverIO Sauce service automatically sets test labels and results. Now that you have run your test on could based emulators, this opens up the ability to scale out the automation of your functional tests. Sometimes, an emulator doesn't quite meet all the requirements necessary to ensure confidence in app quality due to real sensors, older devices, and other aspects like memory and CPU of a real device.
+
+## Running tests on real devices using a cloud-based service
+
+Sauce Labs also provides a real device testing solution through the TestObject platform. With some minor changes to the above `wdio.conf`, we can run UI tests on real devices.
+
+First, create a project in the [TestObject dashboard](https://app.testobject.com/) and upload your `.ipa` or `.apk` file.
+
+There is no official WebDriverIO TestObject service and at the time of writing this blog, Sauce Labs and TestObject do not share the same API calls.
+
+**Update `wdio.conf` file:**
+
+```diff
+- services: ['sauce'],
++ protocol: 'https',
++ host: 'us1.appium.testobject.com',
+- port: '80',
++ port: '443',
++ path: '/api/appium/wd/hub',
+capabilities: {
+  ...
+-  app: 'sauce-storage:myapp.zip'
++  testobject_api_key: <TESTOBJECT_ACCESS_KEY>,
++  testobject_app_id: <APP_ID>,
+}
+```
+
+Refer to [Appium Capabilities for Real Device Testing](https://wiki.saucelabs.com/display/DOCS/Appium+Capabilities+for+Real+Device+Testing) to update your capabilities if needed.
+
+**Run the test:**
+Let's run the test with the updated configuration.
+
+```console
+wdio wdio.conf
+```
+
+Check the [TestObject Dashboard](https://app.testobject.com/) to make sure the test has run. WebDriverIO does not update test labels or results in TestObject Dashboard. TestObject does not aggregate the test results when running through WebDriverIO, therefore, we have to write some extra code so that your test results are in sync. Here are some references that can help with that: <https://github.com/pizzasaurusrex/TestObject> or <https://gist.github.com/rajapanidepu/0e8c0f89671a8a563a7463f8c1ff0413>
+
+## Making test results consistent and predictable
+
+These functional tests are automated and work across multiple service and framework layers. These services and frameworks are not synchronous and not consistently responsive, therefore, we have to accommodate intermittent failures.
+
+When we started to write the UI tests and run them as part of CICD, we observed that the tests were failing for reasons which were not related to the test code itself. An example symptom of this was that a test would pass in the first run, yet, we would see that same test failing randomly in subsequent runs with no changes made to the test code.
```

Comment:
suggested edit:
 An example ~~symptom~~ of this was ~~that~~ a test would pass in the first run, yet ~~,~~ we would see ~~that~~ **the** same test failing randomly in subsequent runs with no changes made to the test code.